velox/common/caching/AsyncDataCache.cpp

```diff
@@ -292,6 +286,13 @@ void CoalescedLoad::setEndState(LoadState endState) {
 void CacheShard::removeEntry(AsyncDataCacheEntry* entry) {
   std::lock_guard<std::mutex> l(mutex_);
   removeEntryLocked(entry);
+  // After the entry is removed from the hash table, a promise can no longer
+  // be made. It is safe to move the promise and realize it.
+
+  auto promise = entry->movePromise();
```

Comment:
it is better to set promise out of lock to avoid deadlock if some future continuation is inline executed?